### PR TITLE
Validate that layer source analysis exist

### LINKED ISSUE #12549: Validate if layer source exists when layer has changed
If there is any change (PUT request) over the layer endpoint, we should check if the `source` attribute exists in the analysis definition. In not, we should trigger an error.

Related to ghost nodes and [this ticket](https://github.com/CartoDB/cartodb/issues/12548).

cc @juanignaciosl @javitonino

### PR DESCRIPTION
**Pull request by juanignaciosl**
This closes #12549.

### CODE REVIEW

**Comment by juanignaciosl** [comment]
Ok, this is the server side validation for a layer: it won't accept any source if source doesn't exist previously as a visualization analysis.

The first error comes quick. STR:

1. Create a map with a dataset (mine was a simple one with two points, but that shouldn't be relevant).
2. Add a centroid analysis.
3. You get this error

captura de pantalla 2017-08-30 a las 17 41 25

The failing PUT is the edition of the layer, and the next PUT is the analysis change. The first PUT tries to set the new a1 analysis as source, but it doesn't exist yet. It will after the second PUT, which adds it.

@javitonino take a look at the code. Do you think that this is a valid approach? I'll mention xavijam after that so we discuss how to procede.

File: app/models/carto/analysis.rb [hunk 1]
```diff
@@ -84,6 +84,10 @@ def analysis_node
     Carto::AnalysisNode.new(analysis_definition)
   end
 
+  def all_analysis_nodes
+    [analysis_node] + analysis_node.children
```
Comment (javitonino) [inline app/models/carto/analysis.rb]: You probably want `analysis_node.descendants` This only goes down one level.

**Comment by javitonino** [comment]
Approach using validations looks good to me. But we should also add the same validation to widgets `source_id` field and to the analysis model as well (the case when the top node of an analysis is removed before updating the layer should fail).

**Comment by juanignaciosl** [comment]
Ok. I'll continue working on the backend side. Nevertheless, I think that we need a frontender on this to fix [the first problem once you enable validations](https://github.com/CartoDB/cartodb/pull/12689#issuecomment-326033201). I'd work iteratively on this. What do you think, @xavijam , how do you want to proceed?

**Comment by xavijam** [comment]
I feel we already changed that @juanignaciosl, didn't we @rubenmoya ?

**Comment by juanignaciosl** [comment]
I just merged master and the STR are the same.

**Comment by rubenmoya** [comment]
Hi! Sorry for being late to the party.

We did changes related to the requests order in `user-actions`, but I'm not sure if that changes fix this problem, since this is creating a new analysis, and our changes were related to moving analysis out of layers if I'm not mistaken.

**Comment by juanignaciosl** [comment]
@rubenmoya , @xavijam then, if we want to add the validations, we need to fix that. I've just pushed the latest changes for sources validation, including analysis, layers and widget validations (what could go wrong?). I'd need help with the frontend fix for previous STR in order to keep doing the basic manual testing, even before CR.

**Comment by juanignaciosl** [comment]
Closing this. cc @javitonino

**Pull request closed without merging**